Replace transaction set in ConsensusData with hashes

omerfirmak opened the pull request:
Part of #1228

Fixes https://github.com/bpfkorea/agora/issues/993

omerfirmak committed:
TransactionPool: Allow fetching hashes of TXs

omerfirmak committed:
TransactionPool: Let pool be queried with a TX Hash

omerfirmak committed:
Set: Add opApply() @safe overload

omerfirmak committed:
FullNode: Add getTranscations endpoint

With nodes now voting on TX hashes, we need to be able to
fetch TXs from other nodes' TX pools.

omerfirmak committed:
Ledger: Keep track of unknown TXs

When node receive a nomination that it can't fully verify, it will
try to fetch the missing TXs. So that when the same nomination can be
validated if it's ever renominated for that round.

omerfirmak committed:
NetworkManager: Fetch missing TXs as part of the catchup

omerfirmak committed:
Test: Add LocalTransaction test

This test creates a network with no TX gossiping. Expect nodes to
request missing TX from peers and reach consensus.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -733,4 +737,29 @@ public class FullNode : API
+
+        Returns:
+            Transactions corresponding to the requested hashes or
+            Transaction.init for hashes that can't be found in the pool
```
That's not what the code does though, it doesn't return `Transaction.init` but just skips those transactions entirely.

omerfirmak replied:
Oops, copy paste error.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -733,4 +737,29 @@ public class FullNode : API
+        {
+            if (found_txs.length >= MaxBatchTranscationsSent)
+                break;
+            auto tx = this.pool.getTransactionByHash(hash);
```
I feel like we can optimize this in the future. I am pretty sure databases are well optimized for looking up several keys at once, and not just one at a time.

omerfirmak replied:
Agreed, this was just to get stuff working.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -800,6 +800,8 @@ extern(D):
             if (auto fail_reason = this.ledger.validateConsensusData(data))
             {
+                if (fail_reason == Ledger.InvalidConsensusDataReason.MayBeValid)
+                    return ValidationLevel.kInvalidValue;
```
By SCP's definition if you return `kInvalidValue` then it means the node may never vote this value as valid afterwards. Are you sure we're not supposed to use `kMaybeValidValue`?

omerfirmak replied:
Well, initially we returned `kMaybeValidValue` here. But I had some problems in the balloting phase. Node started to confirm statements without having access to all the previous blocks. I completely went from these [two](https://github.com/bpfkorea/agora/blob/bbf523199b6efd1667ee0e3d79ad59d4c3ea93c2/source/agora/consensus/protocol/Nominator.d#L539) [comments](https://github.com/bpfkorea/agora/blob/406a22df4c33dfd2f2e3a61811fe6e73aea94852/source/scpp/src/scp/SCPDriver.h#L56-L58). and decided `kInvalidValue` was more appropriate. In nominating phase it does not matter what we return they are the same. In balloting we should still fully validate the value, but our ledger may be a bit behind to return `kMaybeValidValue`. That is indeed what Stellar does actually. They only return `kMaybeValidValue` [here](https://github.com/stellar/stellar-core/blob/54b03f755ae5d5aa12a799c8f1ee4d87fc9d1a1d/src/herder/HerderSCPDriver.cpp#L237), which is the case where the value is fully validated but ledger is not fully caught up.

omerfirmak replied:
> then it means the node may never vote this value as valid afterwards

I am not entirely sure, but if the same value is renominated, it should probably count as a new value.

AndrejMitrovic replied:
Well okay. The docs are not amazing..

AndrejMitrovic replied:
I mean the SCP docs.

omerfirmak committed:
Build: Exclude network tests from "client" configuration

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -677,15 +685,19 @@ public class Ledger
         const expect_height = Height(this.getBlockHeight() + 1);
         auto utxo_finder = this.utxo_set.getUTXOFinder();
-        auto unknown_txs = false;
+        auto missing_ctx = false;
```
Is "ctx" supposed to mean "context"?

omerfirmak replied:
Yep

AndrejMitrovic commented on source/agora/network/NetworkManager.d:
```diff
@@ -893,6 +895,37 @@ public class NetworkManager
+            {
+                try
+                {
+                    ledger.acceptTransaction(tx);
```
You need to check the return value.

omerfirmak replied:
Do we want to log it, if its invalid? I could not think of any other reason for checking if it is valid or not.

AndrejMitrovic replied:
You can't think of any? Check the next line.

If the tx is invalid then the missing tx set should not be modified..

AndrejMitrovic replied:
Wait I just realized something.

AndrejMitrovic replied:
My assumption was wrong. In fact if the tx is invalid there is no other tx with a matching hash, that would make it a hash collision.

AndrejMitrovic replied:
I wonder if this can be abused in some way. If a bad actor nominates a big tx set of invalid hashes, will that cause the node to constantly try to download these txes?

AndrejMitrovic replied:
by invalid I mean fake hashes of non-existing txes.

omerfirmak replied:
It was hard to follow but, yes :)

omerfirmak replied:
Oops I didnt see the last 2 messages when I replied to you.
 
 > I wonder if this can be abused in some way. If a bad actor nominates a big tx set of invalid hashes, will that cause the node to constantly try to download these txes?
 
Until the end of that round, yes. We clear the unknown txs at the end of each round. "big tx set" will probably be bound by some upper limit too eventually. Maybe the block size or tx count?

omerfirmak replied:
We can maybe give up on looking for hashes after a certain amount of tries

AndrejMitrovic replied:
Anyway it shouldn't block the PR.

AndrejMitrovic reviewed (commented):
Overall LGTM. Just a few issues.

omerfirmak commented:
I rebased the fee dıstrıbutıon commits and added some changes to better handle the externalization with missing TXs.
Tests pass on my local machine but I feel like block signatures could be easily broken by this. @hewison-chris, can you check if these make sense to you, especially the early returns in the last commit .

The last commit will be formatted/documented and mostly squashed to previous commits.

hewison-chris commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -860,9 +870,14 @@ extern(D):
+
+            Transaction[] externalized_tx_set;
+            if (this.ledger.getValidTXSet(data, externalized_tx_set))
+                return;
```
I feel like this should be an assert(0) here as we can not ignore the invalid tx set when we should be externalizing the block.

omerfirmak replied:
Since we now externalize TX hashes, we might not have all the TXs in our pool at the point the value was externalized. So that could fail for missing TXs as well. In that case we should be able to catch up to the network and sign the unsigned blocks along the way.

hewison-chris replied:
So will this externalize be triggered again after the missing TXs are received?

omerfirmak replied:
No, node will have to download the full block from the network.

hewison-chris replied:
Perhaps I should also not assert in that case, when there are not enough signatures. The node can also wait to recieve the block during a periodic catchup and test again for required number of signatures before externalizing.

hewison-chris replied:
I created [PR #1517](https://github.com/bpfkorea/agora/pull/1517) to address this for block sigs. I agree with your logic here.

omerfirmak replied:
Great, thanks!

omerfirmak committed:
Consensus: Replace TXs in ConsensusData with their hashes

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -529,8 +529,14 @@ extern(D):
-                con_data.timestamp, con_data.enrolls, random_seed, con_data.missing_validators);
+
+            Transaction[] received_tx_set;
+            if (this.ledger.getValidTXSet(con_data, received_tx_set))
```
So it returns `true` if it fails?

AndrejMitrovic replied:
Ah wait it returns a string.

I thought this behavior was deprecated.

AndrejMitrovic replied:
This would make it clearer:

`if (auto error = this.ledger.getValidTXSet(con_data, received_tx_set))`

Although you could argue that the string is unused anyway.

omerfirmak replied:
Yeah, lets just make it explicit.

hewison-chris replied:
If the string is not used then it may as well be a bool. It is not an `isInvalid...` function so it is not obvious what is success status.

AndrejMitrovic commented:
I'll let @hewison-chris finish his review & then we can merge.

hewison-chris commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -529,8 +529,14 @@ extern(D):
+
+            Transaction[] received_tx_set;
+            if (auto fail_reason = this.ledger.getValidTXSet(con_data, received_tx_set))
+                return; // We dont have all the TXs for this block. Try to catchup
```
We probably should log here the fail_reason as there is nothing logged in any of the called functions.

hewison-chris commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -856,8 +868,15 @@ extern(D):
+            Transaction[] externalized_tx_set;
+            if (auto fail_reason = this.ledger.getValidTXSet(data,
+                externalized_tx_set))
+                return;
```
Also here

hewison-chris reviewed (commented):
Looks good, only a couple of logging requests.

linked0 reviewed (approved):
LGTM

omerfirmak commented:
Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.

AndrejMitrovic commented:
> Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.

It would be good to test this in our test environment. It's a significant change.

omerfirmak commented:
> > Thanks for the reviews everyone. I have made the changes and rebased. I think this is good to merge it once it goes green.
> 
> It would be good to test this in our test environment. It's a significant change.

Together with the Coinbase TXs they make a dangerous duo. Once we get `faucet` to include fees we should be able to test extensively.

omerfirmak commented:
@hewison-chris , is this the failure we were seeing in NetworkManager tests?
https://github.com/bpfkorea/agora/pull/1495/checks?check_run_id=1739937365

hewison-chris commented:
I don't see any new logging. Did you think it will be too much logging?

omerfirmak commented:
> I don't see any new logging. Did you think it will be too much logging?

Check L536 and L880. I've added only to the places you commented on.

hewison-chris commented:
> @hewison-chris , is this the failure we were seeing in NetworkManager tests?
> https://github.com/bpfkorea/agora/pull/1495/checks?check_run_id=1739937365

Sorry, looks like the new network unit test for PeriodicCatchup can sometimes fail. Will take a look now.

hewison-chris commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -711,8 +721,14 @@ extern(D):
+
+        Transaction[] signed_tx_set;
+        if (auto fail_reason = this.ledger.getValidTXSet(con_data, signed_tx_set))
+            return;
```
Also needed here.

hewison-chris commented on source/agora/node/Ledger.d:
```diff
@@ -237,8 +240,13 @@ public class Ledger
-            random_seed, data.missing_validators);
+        Transaction[] externalized_tx_set;
+        if (auto fail_reason = this.getValidTXSet(data, externalized_tx_set))
+            return false;
```
Also here

hewison-chris commented on source/agora/node/Ledger.d:
```diff
@@ -254,8 +262,14 @@ public class Ledger
+
+        Transaction[] externalized_tx_set;
+        if (auto fail_reason = this.getValidTXSet(data, externalized_tx_set))
+            return false;
```
Also here

hewison-chris reviewed (approved):
LGTM. Just a couple more log lines.

omerfirmak commented:
Added more logs. 

`PeriodicCatchup` is still failing though. Should we wait for you to take a look @hewison-chris ?

hewison-chris commented:
> Added more logs.
> 
> `PeriodicCatchup` is still failing though. Should we wait for you to take a look @hewison-chris ?

Yes I know why it fails. Fixing it now. I will have to gossip the signatures also after receiving block during catchup.
